spring-integration-core/src/main/java/org/springframework/integration/support/json/EmbeddedJsonHeadersMessageMapper.java

```diff
@@ -0,0 +1,255 @@
+/*
+ * Copyright 2017 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.integration.support.json;
+
+import java.io.UnsupportedEncodingException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.regex.Pattern;
+
+import org.springframework.integration.mapping.BytesMessageMapper;
+import org.springframework.integration.support.MutableMessage;
+import org.springframework.integration.support.MutableMessageHeaders;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessageHeaders;
+import org.springframework.messaging.support.GenericMessage;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+/**
+ * For outbound messages, uses a message-aware Jackson object mapper to render the message
+ * as JSON. For messages with {@code byte[]} payloads, if rendered as JSON, Jackson
+ * performs Base64 conversion on the bytes. If the {@link #setRawBytes(boolean) rawBytes}
+ * property is true (default), the result has the form
+ * &lt;headersLen&gt;&lt;headers&gt;&lt;payloadLen&gt;&lt;payload&gt;; with the headers
+ * rendered in JSON and the payload unchanged.
+ * <p>
+ * By default, all headers are included; you can provide regular expressions to specify a
+ * subset of headers.
+ * <p>
+ * For inbound messages, in addition to pure JSON and the above format, the Spring Cloud
+ * Stream {@code EmbeddedHeadersMessageConverter} format is also supported:
```

Comment:
We have moved the code from `EmbeddedHeadersMessageConverter` so that other binder implementations can use it. Also, you shouldn't need an instance. A strategy would be a good idea (I can see that being used in SCSt as well), but then the strategy shouldn't depend on an `ObjectMapper` (can be injected in the strategy object)